zero/config.org
* TODO
restart-emacs
vterm
vterm-toggle
keycast-mode-line does not work with telephone-line-mode

* configure emacs from stratch with nix
I use doom emacs as my daily driver. 
~doom profiles sync~

~emacs --profile zero~

* speed run

~SPC m x~ to ~M-x~
turn of keycast mode is
~SPC m x~ ~keycast~
keycast-tab-bar-mode

~SPC x s~ ~C-x C-s~ save-buffer
~SPC x d~  ~C-x C-d~  consult-dir
~project-switch-project~

https://github.com/jwiegley/dot-emacs/blob/master/init.org
https://www.patrickdelliott.com/emacs.d/#org256bd49
https://www.lucacambiaghi.com/vanilla-emacs/readme.html#h:0642C8C0-3B4E-4C72-BD41-40F58FFAB736

#+begin_src elisp
(setq gc-cons-percentage 0.5
      gc-cons-threshold (* 128 1024 1024))
#+end_src


* use-package
use nix for package management, use-package for configuration.
#+begin_src elisp
(eval-and-compile
  (defsubst emacs-path (path)
    (expand-file-name path user-emacs-directory))

  (setq package-enable-at-startup nil
        load-path
        (append (list (emacs-path "use-package"))
                (delete-dups load-path)
                (list (emacs-path "lisp")))))
(require 'use-package)

(setq use-package-verbose init-file-debug
      use-package-expand-minimally (not init-file-debug)
      use-package-compute-statistics t
      debug-on-error init-file-debug)
#+end_src

* global variables
#+begin_src elisp
(defconst emacs-environment (system-name))

(defconst emacs-data-suffix
  (cond ((string= "emacsERC" emacs-environment) "alt")
        ((string-match "emacs2[6789]\\(.+\\)$" emacs-environment)
         (match-string 1 emacs-environment))))

(defconst alternate-emacs (string= emacs-data-suffix "alt"))

(defconst user-data-directory
  (emacs-path (if emacs-data-suffix
                  (format "data-%s" emacs-data-suffix)
                "data")))

(defun user-data (dir)
  (expand-file-name dir user-data-directory))
  #+end_src

* Meow
https://github.com/meow-edit/meow

it's not a good idea to have a complex leader keymap
here we create bindings for necessary, high frequency commands

https://github.com/DogLooksGood/dogEmacs/blob/master/elisp/init-modal.el
#+begin_src elisp
    (defun meow-setup ()
      (setq meow-cheatsheet-layout meow-cheatsheet-layout-qwerty)
      (meow-motion-overwrite-define-key
       '("j" . meow-next)
       '("k" . meow-prev)
       '("<escape>" . ignore))
      (meow-leader-define-key
       ;; SPC j/k will run the original command in MOTION state.
       '("j" . "H-j")
       '("k" . "H-k")
       ;; Use SPC (0-9) for digit arguments.
       '("1" . meow-digit-argument)
       '("2" . meow-digit-argument)
       '("3" . meow-digit-argument)
       '("4" . meow-digit-argument)
       '("5" . meow-digit-argument)
       '("6" . meow-digit-argument)
       '("7" . meow-digit-argument)
       '("8" . meow-digit-argument)
       '("9" . meow-digit-argument)
       '("0" . meow-digit-argument)
       '("/" . meow-keypad-describe-key)
       '("?" . meow-cheatsheet)
        ;; high frequency keybindings
   ;; '("e" . "C-x C-e")
   ;; '(")" . "C-)")
   ;; '("}" . "C-}")
   ;; '("," . xref-pop-marker-stack)
   ;; '("." . xref-find-definitions)
      ;; window management
   '("w" . other-window)
   '("W" . window-swap-states)
   '("o" . delete-other-windows)
   '("s" . split-window-right)
   '("-" . split-window-below)
   ;; overwrited motion key
   ;; '("$" . "H-$")
   ;; high frequency commands
   ;; '(";" . comment-dwim)
   '("k" . kill-this-buffer)
   '("d" . consult-dir)
   '("b" . switch-to-buffer)
   ;; '("r" . rg-project)
   '("f" . consult-find)
   '("i" . imenu)
   ;; '("a" . execute-extended-command)
   ;; '("=" . smerge-basic-map)
   ;; '("p" . project-find-file)
   ;; '("j" . project-switch-to-buffer)
   ;; '("t" . tab-bar-switch-to-tab)
     ;; '("l" . project-switch-project)
   ;; '("y" . magit)
   ;; '("n" . org-roam-keymap)
   ;; toggles
    ;; '("L" . display-line-numbers-mode)
   ;; '("S" . smartparens-strict-mode)
   ;; '("T" . telega)
   ;; '("P" . pass)
   ;; '("A" . org-agenda)
   ;; '("D" . docker)
   ;; '("E" . elfeed)
   ;; '("F" . flymake-mode)
   ;; '("\\" . dired-sidebar-toggle-sidebar)
   ;; '("#" . +project-previous-buffer)
   )
      (meow-normal-define-key
       '("0" . meow-expand-0)
       '("9" . meow-expand-9)
       '("8" . meow-expand-8)
       '("7" . meow-expand-7)
       '("6" . meow-expand-6)
       '("5" . meow-expand-5)
       '("4" . meow-expand-4)
       '("3" . meow-expand-3)
       '("2" . meow-expand-2)
       '("1" . meow-expand-1)
       '("-" . negative-argument)
       '(";" . meow-reverse)
       '("," . meow-inner-of-thing)
       '("." . meow-bounds-of-thing)
       '("[" . meow-beginning-of-thing)
       '("]" . meow-end-of-thing)
       '("a" . meow-append)
       '("A" . meow-open-below)
       '("b" . meow-back-word)
       '("B" . meow-back-symbol)
       '("c" . meow-change)
       '("d" . meow-delete)
       '("D" . meow-backward-delete)
       '("e" . meow-next-word)
       '("E" . meow-next-symbol)
       '("f" . meow-find)
       '("g" . meow-cancel-selection)
       '("G" . meow-grab)
       '("h" . meow-left)
       '("H" . meow-left-expand)
       '("i" . meow-insert)
       '("I" . meow-open-above)
       '("j" . meow-next)
       '("J" . meow-next-expand)
       '("k" . meow-prev)
       '("K" . meow-prev-expand)
       '("l" . meow-right)
       '("L" . meow-right-expand)
       '("m" . meow-join)
       '("n" . meow-search)
       '("o" . meow-block)
       '("O" . meow-to-block)
       '("p" . meow-yank)
       '("q" . meow-quit)
       '("Q" . meow-goto-line)
       '("r" . meow-replace)
       '("R" . meow-swap-grab)
       '("s" . meow-kill)
       '("t" . meow-till)
       '("u" . meow-undo)
       '("U" . meow-undo-in-selection)
       '("v" . meow-visit)
       '("w" . meow-mark-word)
       '("W" . meow-mark-symbol)
       '("x" . meow-line)
       '("X" . meow-goto-line)
       '("y" . meow-save)
       '("Y" . meow-sync-grab)
       '("z" . meow-pop-selection)
       '("'" . repeat)
       '("<escape>" . ignore)))
    (use-package meow
      :demand t
      :config
      (meow-setup)
      (meow-global-mode 1))

  (use-package emacs
      :custom
      (tool-bar-mode nil)             ; Hide the outdated icons
      (scroll-bar-mode nil)           ; Hide the always-visible scrollbar
      (inhibit-splash-screen t) ; Remove the "Welcome to GNU Emacs" splash screen
      (use-file-dialog nil)      ; Ask for textual confirmation instead of GUI
      (menu-bar-mode nil)
      (tool-bar-mode nil)
      (scroll-bar-mode nil)
      :config (setq ring-bell-function #'ignore))
#+end_src
* sane default
In confirmation dialogues, we want to be able to type y and n instead of having to spell the whole words:


Make everything use UTF-8:
#+begin_src elisp
(use-package emacs
  :init
  (defalias 'yes-or-no-p 'y-or-n-p))

(use-package emacs
  :init
  (set-charset-priority 'unicode)
  (setq locale-coding-system 'utf-8
        coding-system-for-read 'utf-8
        coding-system-for-write 'utf-8)
  (set-terminal-coding-system 'utf-8)
  (set-keyboard-coding-system 'utf-8)
  (set-selection-coding-system 'utf-8)
  (prefer-coding-system 'utf-8)
  (setq default-process-coding-system '(utf-8-unix . utf-8-unix)))

(use-package emacs
  :init
  (setq-default indent-tabs-mode nil)
  (setq-default tab-width 2))

(use-package emacs
  :init
	(when (eq system-type 'darwin)
		(setq mac-command-modifier 'super)
		(setq mac-option-modifier 'meta)
		(setq mac-control-modifier 'control)))

#+end_src
* dired
#+begin_src elisp
(use-package dired
   :commands dired-jump
  :diminish dired-omit-mode
   :hook
   (dired-mode . dired-hide-details-mode)
   (dired-mode . dired-omit-mode)
  :custom
  (dired-omit-files "\\`[.]?#\\|\\`[.][.]?\\'\\|^\\.DS_Store\\'\\|^\\.project\\(?:ile\\)?\\'\\|^\\.\\(?:svn\\|git\\)\\'\\|^\\.ccls-cache\\'\\|\\(?:\\.js\\)?\\.meta\\'\\|\\.\\(?:elc\\|o\\|pyo\\|swp\\|class\\)\\'")
  (dired-dwim-target t)
  :config

  )

#+end_src

* recentf
#+begin_src elisp
(use-package recentf
  :demand t
  :commands (recentf-mode
             recentf-add-file
             recentf-apply-filename-handlers)
  :custom
  (recentf-auto-cleanup 60)
  (recentf-exclude
   '("~\\'" "\\`out\\'" "\\.log\\'" "^/[^/]*:" "\\.el\\.gz\\'"))
  (recentf-max-saved-items 2000)
  (recentf-save-file (user-data "recentf"))
  :preface
  (defun recentf-add-dired-directory ()
    "Add directories visit by dired into recentf."
    (if (and dired-directory
             (file-directory-p dired-directory)
             (not (string= "/" dired-directory)))
        (let ((last-idx (1- (length dired-directory))))
          (recentf-add-file
           (if (= ?/ (aref dired-directory last-idx))
               (substring dired-directory 0 last-idx)
             dired-directory)))))
  :hook (dired-mode . recentf-add-dired-directory)
  :config
  (recentf-mode 1))

#+end_src

* keycast
#+begin_src elisp
(use-package keycast
  :commands keycast-mode
   )
#+end_src

* Which key
#+begin_src elisp
(use-package which-key
  :demand t
  :diminish
  :config
  (setq which-key-side-window-location 'bottom
	  which-key-sort-order #'which-key-key-order-alpha
	  which-key-sort-uppercase-first nil
	  which-key-add-column-padding 1
	  which-key-max-display-columns nil
	  which-key-side-window-slot -10
	  which-key-side-window-max-height 0.25
	  which-key-idle-delay 0.8
	  which-key-max-description-length 25
	  which-key-allow-imprecise-window-fit t
	  which-key-separator " → " )

    (which-key-mode)
  )

#+end_src


* Font and theme

#+begin_src elisp
(set-face-attribute 'default nil
  :font "PragmataPro Mono Liga"
  :height 180
  :weight 'medium)

(use-package doom-themes
  :demand
  :config
  (load-theme 'doom-palenight t))

(use-package telephone-line
  :config

  (setq telephone-line-primary-left-separator 'telephone-line-cubed-left
      telephone-line-secondary-left-separator 'telephone-line-cubed-hollow-left
      telephone-line-primary-right-separator 'telephone-line-cubed-right
      telephone-line-secondary-right-separator 'telephone-line-cubed-hollow-right)
(setq telephone-line-height 24)
(setq telephone-line-evil-use-short-tag t)
(telephone-line-defsegment* telephone-line-simpler-major-mode-segment ()
  (concat "["
          (if (listp mode-name)
              (car mode-name)
            mode-name)
          "]"))

(telephone-line-defsegment* telephone-line-simple-pos-segment ()
  (concat "%c : " "%l/" (number-to-string (count-lines (point-min) (point-max)))))

(setq telephone-line-lhs
      '((nil . (telephone-line-projectile-buffer-segment))
        (accent . (telephone-line-simpler-major-mode-segment))
        (nil . (telephone-line-meow-tag-segment
                telephone-line-misc-info-segment)))
      telephone-line-rhs
      '((nil . (telephone-line-simple-pos-segment))
        (accent . (telephone-line-buffer-modified-segment))))

(telephone-line-mode 1)

)


(use-package nerd-icons)

(defun pixel-scroll-setup ()
  (interactive)
  (setq pixel-scroll-precision-large-scroll-height 1)
  (setq pixel-scroll-precision-interpolation-factor 1))

(when (boundp 'pixel-scroll-precision-mode)
  (pixel-scroll-setup)
  (add-hook 'prog-mode-hook #'pixel-scroll-precision-mode)
  (add-hook 'org-mode-hook #'pixel-scroll-precision-mode))
#+end_src

* orderless
#+begin_src elisp
(use-package orderless
  :demand t
  :custom
  (completion-styles '(orderless basic))
  (completion-category-overrides
   '((file (styles basic partial-completion)))))

#+end_src
* embark
https://github.com/oantolin/embark
#+begin_src elisp
(use-package marginalia
  :config
  (marginalia-mode))

(use-package embark

  :bind
  (("C-." . embark-act)         ;; pick some comfortable binding
   ("C-;" . embark-dwim)        ;; good alternative: M-.
   ("C-h B" . embark-bindings)) ;; alternative for `describe-bindings'

  :init

  ;; Optionally replace the key help with a completing-read interface
  (setq prefix-help-command #'embark-prefix-help-command)

  ;; Show the Embark target at point via Eldoc. You may adjust the
  ;; Eldoc strategy, if you want to see the documentation from
  ;; multiple providers. Beware that using this can be a little
  ;; jarring since the message shown in the minibuffer can be more
  ;; than one line, causing the modeline to move up and down:

  ;; (add-hook 'eldoc-documentation-functions #'embark-eldoc-first-target)
  ;; (setq eldoc-documentation-strategy #'eldoc-documentation-compose-eagerly)

  :config

  ;; Hide the mode line of the Embark live/completions buffers
  (add-to-list 'display-buffer-alist
               '("\\`\\*Embark Collect \\(Live\\|Completions\\)\\*"
                 nil
                 (window-parameters (mode-line-format . none)))))

;; Consult users will also want the embark-consult package.
(use-package embark-consult
  :ensure t ; only need to install it, embark loads it after consult if found
  :hook
  (embark-collect-mode . consult-preview-at-point-mode))

#+end_src

* corfu and cape
https://github.com/minad/corfu

https://github.com/minad/cape
#+begin_src elisp
(use-package corfu
  ;; Optional customizations
  ;; :custom
  ;; (corfu-cycle t)                ;; Enable cycling for `corfu-next/previous'
  ;; (corfu-auto t)                 ;; Enable auto completion
  ;; (corfu-separator ?\s)          ;; Orderless field separator
  ;; (corfu-quit-at-boundary nil)   ;; Never quit at completion boundary
  ;; (corfu-quit-no-match nil)      ;; Never quit, even if there is no match
  ;; (corfu-preview-current nil)    ;; Disable current candidate preview
  ;; (corfu-preselect 'prompt)      ;; Preselect the prompt
  ;; (corfu-on-exact-match nil)     ;; Configure handling of exact matches
  ;; (corfu-scroll-margin 5)        ;; Use scroll margin

  ;; Enable Corfu only for certain modes.
  ;; :hook ((prog-mode . corfu-mode)
  ;;        (shell-mode . corfu-mode)
  ;;        (eshell-mode . corfu-mode))

  ;; Recommended: Enable Corfu globally.  This is recommended since Dabbrev can
  ;; be used globally (M-/).  See also the customization variable
  ;; `global-corfu-modes' to exclude certain modes.
  :config
  (global-corfu-mode))

;; A few more useful configurations...
(use-package emacs
  :init
  ;; TAB cycle if there are only few candidates
  (setq completion-cycle-threshold 3)

  ;; Emacs 28: Hide commands in M-x which do not apply to the current mode.
  ;; Corfu commands are hidden, since they are not supposed to be used via M-x.
  ;; (setq read-extended-command-predicate
  ;;       #'command-completion-default-include-p)

  ;; Enable indentation+completion using the TAB key.
  ;; `completion-at-point' is often bound to M-TAB.
  (setq tab-always-indent 'complete))

;; Add extensions
(use-package cape
  ;; Bind dedicated completion commands
  ;; Alternative prefix keys: C-c p, M-p, M-+, ...
  :bind (("C-c p p" . completion-at-point) ;; capf
         ("C-c p t" . complete-tag)        ;; etags
         ("C-c p d" . cape-dabbrev)        ;; or dabbrev-completion
         ("C-c p h" . cape-history)
         ("C-c p f" . cape-file)
         ("C-c p k" . cape-keyword)
         ("C-c p s" . cape-elisp-symbol)
         ("C-c p e" . cape-elisp-block)
         ("C-c p a" . cape-abbrev)
         ("C-c p l" . cape-line)
         ("C-c p w" . cape-dict)
         ("C-c p :" . cape-emoji)
         ("C-c p \\" . cape-tex)
         ("C-c p _" . cape-tex)
         ("C-c p ^" . cape-tex)
         ("C-c p &" . cape-sgml)
         ("C-c p r" . cape-rfc1345))
  :init
  ;; Add to the global default value of `completion-at-point-functions' which is
  ;; used by `completion-at-point'.  The order of the functions matters, the
  ;; first function returning a result wins.  Note that the list of buffer-local
  ;; completion functions takes precedence over the global list.
  (add-to-list 'completion-at-point-functions #'cape-dabbrev)
  (add-to-list 'completion-at-point-functions #'cape-file)
  (add-to-list 'completion-at-point-functions #'cape-elisp-block)
  ;;(add-to-list 'completion-at-point-functions #'cape-history)
  ;;(add-to-list 'completion-at-point-functions #'cape-keyword)
  ;;(add-to-list 'completion-at-point-functions #'cape-tex)
  ;;(add-to-list 'completion-at-point-functions #'cape-sgml)
  ;;(add-to-list 'completion-at-point-functions #'cape-rfc1345)
  ;;(add-to-list 'completion-at-point-functions #'cape-abbrev)
  ;;(add-to-list 'completion-at-point-functions #'cape-dict)
  ;;(add-to-list 'completion-at-point-functions #'cape-elisp-symbol)
  ;;(add-to-list 'completion-at-point-functions #'cape-line)
)
#+end_src

* consult
https://github.com/minad/consult

#+begin_src elisp
(use-package consult
  ;; Replace bindings. Lazily loaded due by `use-package'.
  :bind  (([remap repeat-complex-command] . consult-complex-command)
         ([remap switch-to-buffer] . consult-buffer)
         ([remap switch-to-buffer-other-window] . consult-buffer-other-window)
         ([remap switch-to-buffer-other-frame] . consult-buffer-other-frame)
         ([remap project-switch-to-buffer] . consult-project-buffer)
         ([remap bookmark-jump] . consult-bookmark)
         ([remap goto-line] . consult-line)
         ;; ([remap imenu] . consult-imenu)
         ([remap yank-pop] . consult-yank-pop)
         ("C-c M-x" . consult-mode-command)
         ("C-c h"   . consult-history)
         ("C-c K"   . consult-kmacro)
         ;; ("C-c i"   . consult-info)
         )
  ;; Enable automatic preview at point in the *Completions* buffer. This is
  ;; relevant when you use the default completion UI.
  :hook (completion-list-mode . consult-preview-at-point-mode)

  :custom
  ;; (consult-preview-key "M-i")
  (consult-narrow-key "<")

  :custom-face
  (consult-file ((t (:inherit font-lock-string-face))))

  :functions
  (consult-register-format
   consult-register-window
   consult-xref)

  ;; The :init configuration is always executed (Not lazy)
  :init

  ;; Optionally configure the register formatting. This improves the register
  ;; preview for `consult-register', `consult-register-load',
  ;; `consult-register-store' and the Emacs built-ins.
  (setq register-preview-delay 0.5
        register-preview-function #'consult-register-format)

  ;; Optionally tweak the register preview window.
  ;; This adds thin lines, sorting and hides the mode line of the window.
  (advice-add #'register-preview :override #'consult-register-window)

  ;; Use Consult to select xref locations with preview
  (setq xref-show-xrefs-function #'consult-xref
        xref-show-definitions-function #'consult-xref)

  ;; Configure other variables and modes in the :config section,
  ;; after lazily loading the package.
  :config
  (use-package consult-xref)

  (consult-customize
   consult-theme
   :preview-key '(:debounce 0.2 any)
   consult-ripgrep
   consult-git-grep
   consult-grep
   consult-bookmark
   consult-recent-file
   consult-xref
   consult--source-bookmark
   consult--source-file-register
   consult--source-recent-file
   consult--source-project-recent-file
   :preview-key '(:debounce 0.4 any))

  )
  #+end_src

* consult-dir

#+begin_src elisp
(use-package consult-dir
  :ensure t
  :bind (("C-x C-d" . consult-dir)
         :map vertico-map
         ("C-x C-d" . consult-dir)
         ("C-x C-j" . consult-dir-jump-file)))

#+end_src
* vertico
https://github.com/minad/vertico
#+begin_src elisp
;; Enable vertico
(use-package vertico
  :config
  (vertico-mode)

  ;; Different scroll margin
  ;; (setq vertico-scroll-margin 0)

  ;; Show more candidates
  ;; (setq vertico-count 20)

  ;; Grow and shrink the Vertico minibuffer
  ;; (setq vertico-resize t)

  ;; Optionally enable cycling for `vertico-next' and `vertico-previous'.
  ;; (setq vertico-cycle t)
  )

;; Persist history over Emacs restarts. Vertico sorts by history position.
(use-package savehist
  :config
  (savehist-mode))

;; A few more useful configurations...
(use-package emacs
  :init
  ;; Add prompt indicator to `completing-read-multiple'.
  ;; We display [CRM<separator>], e.g., [CRM,] if the separator is a comma.
  (defun crm-indicator (args)
    (cons (format "[CRM%s] %s"
                  (replace-regexp-in-string
                   "\\`\\[.*?]\\*\\|\\[.*?]\\*\\'" ""
                   crm-separator)
                  (car args))
          (cdr args)))
  (advice-add #'completing-read-multiple :filter-args #'crm-indicator)

  ;; Do not allow the cursor in the minibuffer prompt
  (setq minibuffer-prompt-properties
        '(read-only t cursor-intangible t face minibuffer-prompt))
  (add-hook 'minibuffer-setup-hook #'cursor-intangible-mode)

  ;; Emacs 28: Hide commands in M-x which do not work in the current mode.
  ;; Vertico commands are hidden in normal buffers.
  ;; (setq read-extended-command-predicate
  ;;       #'command-completion-default-include-p)

  ;; Enable recursive minibuffers
  (setq enable-recursive-minibuffers t))

#+end_src


https://gitlab.com/skybert/my-little-friends/-/blob/master/emacs/.emacs
#+begin_src elisp
;; Minimising & quitting Emacs way too many times without wanting to.
(global-unset-key "\C-z")
(global-unset-key "\C-x\C-c")


#+end_src

* ace window

  Quickly switch between open buffer windows
#+begin_src elisp
(use-package ace-window
  :bind
  ("M-o" . ace-window)
  :config
  (setq aw-keys '(?a ?s ?d ?f ?g ?h ?j ?k ?l)
        aw-dispatch-always t))


#+end_src

* avy
#+begin_src elisp
(use-package avy
  :bind ("C-." . avy-goto-char-timer)
  :custom
  (avy-case-fold-search t)
  (avy-timeout-seconds 0.3)
)

#+end_src

* magit
#+begin_src elisp
(require 'magit)
(use-package eglot
  :config
 (add-to-list 'eglot-server-programs
              `(java-mode "/opt/homebrew/bin/jdtls"
                           "-Djava.format.settings.url=https://raw.githubusercontent.com/google/styleguide/gh-pages/eclipse-java-google-style.xml"
                           "-Djava.format.settings.profile=GoogleStyle"
                           ,(concat "--jvm-arg=-javaagent:" (expand-file-name "/Users/yuanwang/Downloads/lombok.jar"))))

  )

#+end_src

* consult-eglot

#+begin_src elisp
  (use-package consult-eglot
  :after (consult eglot)
  :bind (:map eglot-mode-map ("M-g s" . consult-eglot-symbols)))
  #+end_src

* jinx

https://github.com/minad/jinx

#+begin_src elisp
(use-package jinx
  :hook (emacs-startup . global-jinx-mode)
  :bind (("M-$" . jinx-correct)
         ("C-M-$" . jinx-languages)))

#+end_src

* nix
#+begin_src elisp
(use-package nix-mode
  :mode "\\.nix\\'"
  :custom
  (nix-indent-function 'nix-indent-line))

#+end_src

* direnv

#+begin_src elisp
    (use-package direnv
    )
#+end_src

* haskell mode

#+begin_src elisp

(use-package haskell-mode
  :mode (("\\.hs\\(c\\|-boot\\)?\\'" . haskell-mode)
         ("\\.lhs\\'" . haskell-literate-mode)
         ("\\.cabal\\'" . haskell-cabal-mode))
  :bind (:map
         haskell-mode-map
         ("C-c C-h" . my-haskell-hoogle)
         ("C-c C-," . haskell-navigate-imports)
         ("C-c C-." . haskell-mode-format-imports)
         ("C-c C-u" . my-haskell-insert-undefined)
         ("C-c C-z" . haskell-interactive-switch)
         ("M-s")
         ("M-t"))
  :hook

modules/editor/doom/config/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!
;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets.
;; (setq user-full-name "Yuan Wang"
;;       user-mail-address "yuan.wang@workiva.com")
;; Doom exposes five (optional) variables for controlling fonts in Doom. Here
;; are the three important ones:
;;
;; + `doom-font'
;; + `doom-variable-pitch-font'
;; + `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;;
;; They all accept either a font-spec, font string ("Input Mono-12"), or xlfd
;; font string. You generally only need these two:
;; (setq doom-font (font-spec :family "monospace" :size 12 :weight 'semi-light)
;;       doom-variable-pitch-font (font-spec :family "sans" :size 13))

;; there are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
(setq
 doom-theme 'doom-palenight
 doom-modeline-icon nil
 ;; doom-theme 'doom-nano-dark-theme
 doom-font (font-spec :family "PragmataPro" :size 18))
;; (after! doom-themes
;;   (load-theme 'doom-nano-light t))
;; .doom.d/config.el
(setq  fancy-splash-image  "~/.config/wallpapers/doom.svg")

;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
(setq org-directory "~/org/"
      org-roam-directory (concat org-directory "roam/")
      org-agenda-files (append
                        (file-expand-wildcards (concat org-directory "*.org"))
                        (file-expand-wildcards (concat org-directory "agenda/*.org"))
                        (file-expand-wildcards (concat org-directory "projects/*.org")))
      org-default-notes-file (concat org-directory "agenda/inbox.org")
      +org-capture-notes-file (concat org-directory "agenda/inbox.org")
      +org-capture-todo-file (concat org-directory "agenda/inbox.org")
       org-refile-targets '((+org/opened-buffer-files :maxlevel . 9))
      )
(defun +org/opened-buffer-files ()
  "Return the list of files currently opened in Emacs."
  (delq nil
        (mapcar (lambda (x)
                  (if (and (buffer-file-name x)
                           (string-match "\\.org$"
                                         (buffer-file-name x)))
                      (buffer-file-name x)))
                (buffer-list))))

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type nil
      abbrev-file-name (concat doom-user-dir "abbrevs.el"))
;; Prevents some cases of Emacs flickering.
(add-to-list 'default-frame-alist '(inhibit-double-buffering . t))
(use-package! justl
  :config
  (map! :n "e" 'justl-exec-recipe))




;; (use-package! chatgpt-shell
;;   :requires shell-maker
;;   :config
;;   ;; if you are using the "pass" password manager
;;  ((chatgpt-shell-openai-key
;;       (lambda ()
;;         ;; (auth-source-pass-get 'secret "openai-key") ; alternative using pass support in auth-sources
;;         (nth 0 (process-lines "pass" "show" "openapi"))))
;;   ))
;;
;;; Keybinds

;; (map!

 ;; (:after evil-org

 ;;       :map evil-org-mode-map
 ;;       :n "gk" (cmd! (if (org-on-heading-p)
 ;;                         (org-backward-element)
 ;;                       (evil-previous-visual-line)))
 ;;       :n "gj" (cmd! (if (org-on-heading-p)
 ;;                         (org-forward-element)
 ;;                       (evil-next-visual-line))))

 ;; :o "o" #'evil-inner-symbol

 ;; :leader
 ;; (:prefix "w"
 ;;  "w" #'ace-window
 ;;   )

 ;; (:prefix "i"
 ;;  "o" #'org-insert-todo-heading
 ;;  )
 ;; (:prefix "t"
 ;;          "k" #'keycast-mode
 ;;          )

 ;; )


;;; :ui modeline
;; An evil mode indicator is redundant with cursor shape
;; (advice-add #'doom-modeline-segment--modals :override #'ignore)
(setq org-roam-dailies-capture-templates
      '(("d" "default" entry
         "* %?"
         :target (file+head "%<%Y-%m-%d>.org"
                            "#+title: %<%Y-%m-%d>\n* Tasks to do \n* Journal \n* TIL \n"))))

(setq org-agenda-custom-commands
      `(("d" "Dashboard"
         (
          ;; (agenda "" ((org-deadline-warning-days 3)))
          (todo "TODO"
                ((org-agenda-overriding-header "Work")
                 (org-agenda-files '("~/org/agenda/workiva.org"))
                 (org-agenda-text-search-extra-files nil)))

          (tags-todo "+PRIORITY=\"A\""
                     ((org-agenda-overriding-header "High Priority")))
          (tags-todo "+followup" ((org-agenda-overriding-header "Needs Follow Up")))
          (tags-todo "+next"
                     ((org-agenda-overriding-header "Next Actions")
                      (org-agenda-max-todos nil)))
          (todo "TODO"
                ((org-agenda-overriding-header "Unprocessed Inbox Tasks")
                 (org-agenda-files '("~/org/agenda/inbox.org"))
                 (org-agenda-text-search-extra-files nil )))))

        ("i" "Next Tasks"
         (
          (tags-todo "+next"
                     ((org-agenda-overriding-header "Next Tasks")))))))

(flycheck-define-checker vale
  "A checker for prose"
  :command ("vale" "--output" "line"
            source)
  :standard-input nil
  :error-patterns
  ((error line-start (file-name) ":" line ":" column ":" (id (one-or-more (not (any ":")))) ":" (message) line-end))
  :modes (markdown-mode gfm-mode org-mode text-mode))
 (add-to-list 'flycheck-checkers 'vale 'append)
;; (use-package! interaction-log)
(use-package! protobuf-mode)

(use-package! thrift-mode
  :config
  (add-to-list 'auto-mode-alist '("\\.frugal\\'" . thrift-mode ) ))
(use-package! org-wild-notifier
  :config
  (after! org
    (org-wild-notifier-mode)
    (setq
     org-wild-notifier-alert-time '(60 30)
     alert-default-style 'osx-notifier))
  )
(after! browse-at-remote
  (setq browse-at-remote-add-line-number-if-no-region-selected t))
;; I like abbrevs, at least in org mode
;; (add-hook 'org-mode-hook #'abbrev-mode)
;; (add-transient-hook! 'focus-out-hook (atomic-chrome-start-server))
(after! org
  (setq org-eukleides-path (getenv "EUKLEIDES_PATH")
        org-agenda-dim-blocked-tasks nil
        org-agenda-inhibit-startup t
        org-agenda-use-tag-inheritance nil)
  (add-to-list 'org-modules 'org-habit)
  )

(use-package! super-save
  :config
  (add-to-list 'super-save-triggers 'vertico)
  (add-to-list 'super-save-triggers 'magit)
  (add-to-list 'super-save-triggers 'find-file)
  (add-to-list 'super-save-triggers 'winner-undo)
  (super-save-mode +1)
 )

;; (use-package! chatgpt-shell
;;   :requires shell-maker
;;   )
(use-package! embark-vc
  :after embark)
(use-package! evil-replace-with-register
  :config
  (setq evil-replace-with-register-key (kbd "gr"))
  (evil-replace-with-register-install))


(use-package! keycast
  :commands keycast-mode
  :config
  (define-minor-mode keycast-mode
    "Show current command and its key binding in the mode line."
    :global t
    (if keycast-mode
        (progn
          (add-hook 'pre-command-hook 'keycast--update t)
          (add-to-list 'global-mode-string '("" keycast-mode-line " ")))
      (remove-hook 'pre-command-hook 'keycast--update)
      (setq global-mode-string (remove '("" keycast-mode-line " ") global-mode-string))))
  (custom-set-faces!
    '(keycast-command :inherit doom-modeline-debug
      :height 0.9)
    '(keycast-key :inherit custom-modified
      :height 1.1
      :weight bold))
  )
;; (use-package! vlf-setup
;;   :defer-incrementally  vlf-base vlf-write vlf-search vlf-occur vlf-follow vlf-ediff vlf)
(setq
 magit-list-refs-sortby "-committerdate"
 magit-log-section-commit-count 25
 ;; org-export-with-broken-links t
 ;; org-id-track-globally t
 ;; lsp-modeline-code-actions-enable nil
 ;;  lsp-modeline-diagnostics-enable nil

  )
(after! eglot
  :config
 (add-to-list 'eglot-server-programs
              `(java-mode "jdtls-with-lomok"
                           ))

  )


;; (setq lsp-java-configuration-runtimes '[(:name "JavaSE-17"
;; 						:path "/Users/yuanwang/.sdkman/candidates/java/17.0.5-amzn/"
;; 						:default t)])
;; Here are some additional functions/macros that could help you configure Doom:
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.

(setf org-html-mathjax-template
      "<script src=\"https://polyfill.io/v3/polyfill.min.js?features=es6\"></script><script id=\"MathJax-script\" async src=\"https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js\"></script>"
      )
